### ISSUE TITLE
New rule option to ignore trailing whitespace on blank lines

### ISSUE THREAD

**Issue by gonzofish** [palantir/tslint#481]
When I write code with spaces between lines, such as (each "." is a space):

``` ts
((): void => {
....'use strict';
....
....console.info('hello');
})();
```

The no-trailing-whitespace rule tells me there is an error on the blank line (in this case line 3). That seems wrong since most editors are going to insert spaces by default, this makes the user have to remove spaces for it to pass. This happens with both spaces & tabs.

**Comment by adidahiya** [comment]
This is exactly the intent of the `no-trailing-whitespace` rule. Most modern editors actually do the _opposite_ -- they remove trailing whitespace upon save, including on empty lines like this -- so it feels like best practice to report this as a lint failure.

If you really want to allow this kind of trailing whitespace in your codebase, you can create a custom tslint rule (see docs for this in the readme) in your project, copy the implementation of `noTrailingWhitespaceRule.ts`, and have it ignore empty lines.

**Comment by gonzofish** [comment]
Thanks a lot!

**Comment by aciccarello** [comment]
I would like to see this as an option. Something like `ignore-empty-lines.

**Comment by adidahiya** [comment]
Looks like there's some interest in this option. Low priority for us, but we're open to PRs.

**Comment by code-chris** [comment]
+1

**Comment by oliverjanik** [comment]
VsCode by default does not remove whitespace from empty lines.

**Comment by cwmacdon** [comment]
for VSCode: Version 0.5.0 is out and a new setting has been added to trim whitespace on save (files.trimTrailingWhitespace)

**Comment by OGZCoder** [comment]
A easy fix if your not worried about the white spaces is to pipe your source through gulp-trimlines.

`npm install gulp-trimlines --save-dev`

Then in the gulp task:

`function tslint(done) {
    return gulp.src(PATHS.typescript.app)
        .pipe($.trimlines())
        .pipe($.tslint({
            formatter: 'prose',
            noUnusedParameters: true,
            noUnusedLocals: true
        }))
        .pipe($.tslint.report());
        done();
}`

**Comment by doom777** [comment]
i think it makes sense to have trailing whitespace on empty lines, especially if you are using something that semantically uses indentation, such as YAML or Python. Can we have a force-trailing-whitespace rule?

**Comment by vance** [comment]
@doom777 um, just curious, why would you use tslint on python?

**Comment by doom777** [comment]
well, you might have something embedded, but I was genereally referring to the overall philosophy. A blank line belongs to a method or a class, or an if block, and therefore should have its indentation.

**Comment by aaronbeall** [comment]
+1. WebStorm/IntelliJ IDEA has a format option to "preserve indentation on blank lines" which I always use. To me, the spirit of "trailing whitespace" is not to remove indentation.

**Comment by gabrielizalo** [comment]
+1

**Comment by zartc** [comment]
FYI, ESLint has a "skipBlankLines" option to precisely bypass this problem.

**Comment by adidahiya** [comment]
The issue is marked as "Accepting PRs". Feel free to send one.

**Comment by ajafff** [comment]
Fixed by #3346

**Comment by ebewley** [comment]
I believe blank lines and trailing whitespace ought to be treated separately.  I do not like for my devs to leave trailing whitespace, but I **do** like for them to place **one** blank line before jsdoc comments to improve readability.  I would like to see this split apart, because if we turn off _no-trailing-whitespace_ for blank lines, then it allows any number of blank lines which can throw off diff comparisons, and does nothing to improve readability.  A _no-blank-lines_ rule could then allow for exceptions such as _before-jsdoc_, _after-method_, etc.
